packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -42,332 +33,327 @@ const DEFAULT_QUERY = {
+		// don't want to modify the original term.
+		const term = { ...originalTerm };
+
+		// Map and filter the children, recursive so we deal with grandchildren
+		// and any deeper levels.
+		if ( term.children.length > 0 ) {
+			term.children = term.children
+				.map( matchTermsForFilter )
+				.filter( ( child ) => child );
+		}
+
+		// If the term's name contains the filterValue, or it has children
+		// (i.e. some child matched at some point in the tree) then return it.
+		if (
+			-1 !==
+				term.name.toLowerCase().indexOf( filterValue.toLowerCase() ) ||
+			term.children.length > 0
+		) {
+			return term;
+		}
+
+		// Otherwise, return false. After mapping, the list of terms will need
+		// to have false values filtered out.
+		return false;
+	};
+	return matchTermsForFilter;
+}
+
+/**
+ * Hierarchical term selector.
+ *
+ * @param {Object} props      Component props.
+ * @param {string} props.slug Taxonomy slug.
+ * @return {WPElement}        Hierarchical term selector component.
+ */
+function HierarchicalTermSelector( { slug } ) {
+	const instanceId = useInstanceId( HierarchicalTermSelector );
+	const [ adding, setAdding ] = useState( false );
+	const [ formName, setFormName ] = useState( '' );
+	/**
+	 * @type {[number|'', Function]}
+	 */
+	const [ formParent, setFormParent ] = useState( '' );
+	const [ showForm, setShowForm ] = useState( false );
+	const [ filterValue, setFilterValue ] = useState( '' );
+	const [ filteredTermsTree, setFilteredTermsTree ] = useState( [] );
+	const debouncedSpeak = useDebounce( speak, 500 );
+
+	const {
+		hasCreateAction,
+		hasAssignAction,
+		terms,
+		loading,
+		availableTerms,
+		taxonomy,
+	} = useSelect(
+		( select ) => {
+			const { getCurrentPost, getEditedPostAttribute } = select(
+				editorStore
+			);
+			const { getTaxonomy, getEntityRecords, isResolving } = select(
+				coreStore
+			);
+			const _taxonomy = getTaxonomy( slug );
+
+			return {
+				hasCreateAction: _taxonomy
+					? get(
+							getCurrentPost(),
+							[
+								'_links',
+								'wp:action-create-' + _taxonomy.rest_base,
+							],
+							false
+					  )
+					: false,
+				hasAssignAction: _taxonomy
+					? get(
+							getCurrentPost(),
+							[
+								'_links',
+								'wp:action-assign-' + _taxonomy.rest_base,
+							],
+							false
+					  )
+					: false,
+				terms: _taxonomy
+					? getEditedPostAttribute( _taxonomy.rest_base )
+					: EMPTY_ARRAY,
+				loading: isResolving( 'getEntityRecords', [
+					'taxonomy',
+					slug,
+					DEFAULT_QUERY,
+				] ),
+				availableTerms:
+					getEntityRecords( 'taxonomy', slug, DEFAULT_QUERY ) ||
+					EMPTY_ARRAY,
+				taxonomy: _taxonomy,
+			};
+		},
+		[ slug ]
+	);
+
+	const { editPost } = useDispatch( editorStore );
+	const { saveEntityRecord } = useDispatch( coreStore );
+
+	const availableTermsTree = useMemo(
+		() => sortBySelected( buildTermsTree( availableTerms ), terms ),
+		[ availableTerms ]
```

Comment:
It seems `terms` is missing in the dependencies

Reply:
Adding `term` to deps causes the termTree to be reordered when checking/unchecking categories. 
Is there a better way to deal with this problem?

Related: https://github.com/WordPress/gutenberg/pull/33418#issuecomment-881354021

Reply:
ok, so basically we want the "checked" items first but only on initial load? It seems that's the behavior on trunk?

Reply:
This is also the case for trunk. When a component was mounted, fetchTerms was executed, and sorting was performed within it.  Also, when adding a new term, fetchTerms was executed in the same way.

Reply:
In that case maybe we should clarify with a comment. I think it's probably a behavior we'd want to revisit, it's very random.

Reply:
I wrote a comment, how do you like it?